### PR TITLE
[inductor] Avoid cross-extern fusion when allow_peak_memory_increasing_fusion is off

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #188059
* __->__ #187842
* #188058
* #188057

# [inductor] Avoid cross-extern fusion when allow_peak_memory_increasing_fusion is off

(PR https://github.com/pytorch/pytorch/pull/187842)

## Summary

Add a generic fusion guard for cases where fusing across an extern branch would make another
branch output live across that extern user. The guard is active only when
`allow_peak_memory_increasing_fusion=False`, so default `True` keeps previous behavior;
peak-memory-conscious callers (full-graph GraphTrainer compilation) opt in.

Example: branch A produces a 64 MB output consumed first by an extern `mm`. Branch B produces
a 64 MB output consumed later by pointwise code. Fusing A and B can force B's output to be
materialized before the extern and kept live across it, even though the shared input read
saved by fusion is small. In peak-aware mode this fusion is rejected; with
`allow_peak_memory_increasing_fusion=True` it is still allowed.

## Benchmark (marginal)

GraphTrainer full Inductor, 8-GPU local Rigi, no fuse regions, **default Inductor buffer
reuse ON**, `TORCHINDUCTOR_ALLOW_PEAK_MEMORY_INCREASING_FUSION=0`, `nsteps=10 warmup=3`.
Marginal active / reserved peak vs the parent commit (on top of the disjoint-horizontal
guard):

| config | active before -> after | reserved before -> after |
|---|---|---|
| d2 w2048 v256 (small vocab) | 3691 -> 3107 (**-584 MiB**) | 3934 -> 3278 (-656 MiB) |
| d8 w2048 v256 (small vocab) | 5899 -> 5315 (**-584 MiB**) | 6142 -> 5742 (-400 MiB) |
| d8 w7168 full vocab | 19636 -> 18786 (**-850 MiB**) | 21656 -> 22048 (+392 MiB) |

(Reserved is allocator-cache and noisier; the d8 w7168 +392 MiB is within that noise while
active drops 850 MiB.)

## End-to-end: full 2-commit stack vs default torch.compile

Disjoint-horizontal + cross-extern guards together, GraphTrainer (default reuse ON,
`allow_peak_memory_increasing_fusion=0`) vs default `torch.compile` (reuse ON,
`allow_peak_memory_increasing_fusion=1`). Step time is the median of steady steps.

| config | metric | torch.compile | GraphTrainer + stack | delta |
|---|---|---:|---:|---:|
| d2 w2048 v256 | active MiB | 3845.9 | 3106.2 | **-739.7** |
| | reserved MiB | 4395 | 3278 | -1117 |
| | step ms | 151.0 | 125.5 | -25.5 |
| d8 w2048 v256 | active MiB | 5314.1 | 5315.0 | +0.9 (tie) |
| | reserved MiB | 5834.5 | 5742 | -92.5 |
| | step ms | 354.0 | 353.0 | -1.0 |
| d8 w7168 full vocab | active MiB | 19918.4 | 18785.5 | **-1132.9** |
| | reserved MiB | 21972 | 22048 | +76 (tie) |
| | step ms | 3659.0 | 3663.0 | +4.0 (tie) |

GraphTrainer with the 2-commit stack is not worse than default `torch.compile` on active,
reserved, or step time on any config (wins or ties within noise), and recovers up to ~1.1 GiB
of active peak on the capacity config. Final loss matches `torch.compile` to ~1e-5.

## Test plan

```bash
python -m pytest test/inductor/test_inductor_scheduler.py -k "extern_branch" -v

# Rigi full-graph GraphTrainer reproduction (d8 w7168 full vocab):
cd <rigi>
TORCHINDUCTOR_ALLOW_PEAK_MEMORY_INCREASING_FUSION=0 \
RIGI_RESEARCH_DISABLE_FUSE_REGIONS=all \
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 NCCL_SOCKET_IFNAME=lo \
  bv2/tools/_local_run bv2.train nsteps=10 warmup_nsteps=3 evals=None \
  c.model.txt_unemb.chunksz=4096 c.data.min_nouns=256 c.data.max_nouns=256 \
  c.model.depth=8 c.model.dim=7168 use_graph_trainer=True
```

---
Authored with the assistance of an AI coding assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
lets sync more off line on approach for this. also i know you are unblocked
